File: crates/desktop/src/logging.rs
use once_cell::sync::OnceCell;
use serde_json::{Map, Value, json};
use std::fmt::Display;
use time::OffsetDateTime;
use time::UtcOffset;
use time::format_description::well_known::Rfc3339;

#[derive(Clone, Debug)]
pub struct LoggingContext {
    pub session_id: String,
    pub build_version: String,
    pub build_commit: Option<String>,
    pub platform_os: String,
    pub platform_arch: String,
}

static CONTEXT: OnceCell<LoggingContext> = OnceCell::new();

pub fn init_context(session_id: String) -> Option<&'static LoggingContext> {
    CONTEXT
        .set(LoggingContext {
            session_id,
            build_version: env!("CARGO_PKG_VERSION").to_string(),
            build_commit: option_env!("GIT_COMMIT_HASH").map(|s| s.to_string()),
            platform_os: std::env::consts::OS.to_string(),
            platform_arch: std::env::consts::ARCH.to_string(),
        })
        .ok();
    CONTEXT.get()
}

pub fn context() -> Option<&'static LoggingContext> {
    CONTEXT.get()
}

pub fn log_record(
    severity: u8,
    component: &str,
    event_name: &str,
    message: &str,
    correlation_id: &str,
    metadata: Option<Value>,
    module: &str,
    file: &str,
    line: u32,
) {
    let level = severity_to_level(severity);
    let level_name = level_to_str(level);
    let timestamp = OffsetDateTime::now_utc()
        .format(&Rfc3339)
        .unwrap_or_else(|_| {
            OffsetDateTime::now_utc()
                .format(&Rfc3339)
                .unwrap_or_default()
        });

    let mut record = Map::new();
    record.insert("timestamp".into(), Value::String(timestamp));
    record.insert("level".into(), Value::String(level_name.to_string()));
    record.insert("syslog.severity".into(), Value::from(severity as u8));
    record.insert(
        "syslog.severity_text".into(),
        Value::String(severity_text(severity).to_string()),
    );
    record.insert("message".into(), Value::String(message.to_string()));
    record.insert("component".into(), Value::String(component.to_string()));
    record.insert("event_name".into(), Value::String(event_name.to_string()));
    record.insert(
        "correlation_id".into(),
        Value::String(correlation_id.to_string()),
    );

    if let Some(ctx) = CONTEXT.get() {
        record.insert("session_id".into(), Value::String(ctx.session_id.clone()));
        record.insert(
            "build".into(),
            json!({
                "version": ctx.build_version,
                "commit": ctx.build_commit.as_deref().unwrap_or("unknown"),
            }),
        );
        record.insert(
            "platform".into(),
            json!({
                "os": ctx.platform_os,
                "arch": ctx.platform_arch,
            }),
        );
    } else {
        record.insert("session_id".into(), Value::String("unknown".into()));
    }

    record.insert(
        "source".into(),
        json!({
            "layer": "rust",
            "module": module,
            "file": file,
            "line": line,
        }),
    );

    if let Some(payload) = metadata {
        if let Value::Object(map) = payload {
            for (key, value) in map {
                record.entry(key).or_insert(value);
            }
        }
    }

    let payload = Value::Object(record);
    let encoded = serde_json::to_string(&payload)
        .unwrap_or_else(|_| "{\"message\":\"log serialization failed\"}".to_string());
    log::log!(target: "tauri", level, "{}", encoded);
}

fn severity_to_level(severity: u8) -> log::Level {
    match severity {
        0..=3 => log::Level::Error,
        4 => log::Level::Warn,
        5 | 6 => log::Level::Info,
        7 => log::Level::Debug,
        _ => log::Level::Info,
    }
}

fn level_to_str(level: log::Level) -> &'static str {
    match level {
        log::Level::Error => "ERROR",
        log::Level::Warn => "WARN",
        log::Level::Info => "INFO",
        log::Level::Debug => "DEBUG",
        log::Level::Trace => "TRACE",
    }
}

fn severity_text(severity: u8) -> &'static str {
    match severity {
        0 => "Emergency",
        1 => "Alert",
        2 => "Critical",
        3 => "Error",
        4 => "Warning",
        5 => "Notice",
        6 => "Informational",
        7 => "Debug",
        _ => "Informational",
    }
}

#[derive(serde::Serialize)]
pub struct LoggingContextDto {
    pub session_id: String,
    pub build_version: String,
    pub build_commit: Option<String>,
    pub platform_os: String,
    pub platform_arch: String,
}

pub fn get_logging_context() -> std::result::Result<LoggingContextDto, String> {
    if let Some(ctx) = CONTEXT.get() {
        Ok(LoggingContextDto {
            session_id: ctx.session_id.clone(),
            build_version: ctx.build_version.clone(),
            build_commit: ctx.build_commit.clone(),
            platform_os: ctx.platform_os.clone(),
            platform_arch: ctx.platform_arch.clone(),
        })
    } else {
        Err("logging context not initialised".into())
    }
}

#[macro_export]
macro_rules! log_event {
    (
        severity = $severity:expr,
        component = $component:expr,
        event = $event:expr,
        message = $message:expr $(,
        correlation_id = $correlation:expr)? $(,
        metadata = $metadata:expr)? $(,)?
    ) => {{
        let correlation = $crate::log_event!(@corr $($correlation)?);
        let metadata = $crate::log_event!(@meta $($metadata)?);
        $crate::logging::log_record(
            $severity,
            $component,
            $event,
            $message,
            correlation,
            metadata,
            module_path!(),
            file!(),
            line!(),
        );
    }};
    (@corr $value:expr) => { $value };
    (@corr) => { "unknown" };
    (@meta $value:expr) => { Some($value) };
    (@meta) => { None };
}

File: src/lib/logging.ts
import { getLoggingContext } from '@/adapters/logging-context';

export type Severity =
  | 'emergency'
  | 'alert'
  | 'critical'
  | 'error'
  | 'warning'
  | 'notice'
  | 'info'
  | 'informational'
  | 'debug';

type NormalizedSeverity =
  'emergency' | 'alert' | 'critical' | 'error' | 'warning' | 'notice' | 'informational' | 'debug';

interface SeverityAttributes {
  readonly level: 'ERROR' | 'WARN' | 'INFO' | 'DEBUG';
  readonly severityNumber: number;
  readonly severityText: string;
  readonly consoleMethod: 'error' | 'warn' | 'info' | 'debug';
}

/** @public */
export interface LogSource {
  readonly module?: string;
  readonly function?: string;
}

export interface LogEntry {
  readonly severity: Severity;
  readonly component: string;
  readonly eventName: string;
  readonly message: string;
  readonly correlationId?: string;
  readonly source?: LogSource;
  readonly metadata?: Record<string, unknown>;
}

/**
 * Normalize `info` severity so the downstream sinks receive a stable set of labels.
 * @param severity severity value from the caller.
 */
function normalizeSeverity(severity: Severity): NormalizedSeverity {
  return severity === 'info' ? 'informational' : severity;
}

/**
 * Attributes for every normalized severity level, keyed by label.
 *
 * The `Record<NormalizedSeverity, ...>` type makes the compiler require an
 * entry for every severity, so the mapping stays exhaustive.
 */
const severityAttributesByLevel: Record<NormalizedSeverity, SeverityAttributes> = {
  emergency: {
    level: 'ERROR',
    severityNumber: 0,
    severityText: 'Emergency',
    consoleMethod: 'error',
  },
  alert: { level: 'ERROR', severityNumber: 1, severityText: 'Alert', consoleMethod: 'error' },
  critical: { level: 'ERROR', severityNumber: 2, severityText: 'Critical', consoleMethod: 'error' },
  error: { level: 'ERROR', severityNumber: 3, severityText: 'Error', consoleMethod: 'error' },
  warning: { level: 'WARN', severityNumber: 4, severityText: 'Warning', consoleMethod: 'warn' },
  notice: { level: 'INFO', severityNumber: 5, severityText: 'Notice', consoleMethod: 'info' },
  informational: {
    level: 'INFO',
    severityNumber: 6,
    severityText: 'Informational',
    consoleMethod: 'info',
  },
  debug: { level: 'DEBUG', severityNumber: 7, severityText: 'Debug', consoleMethod: 'debug' },
};

/**
 * Lookup derived from {@link severityAttributesByLevel} so callers can resolve
 * attributes by a runtime severity value without an object index expression.
 */
const severityAttributesLookup = new Map(Object.entries(severityAttributesByLevel));

/**
 * Describe the attributes that every severity level maps to.
 * @param severity normalized severity label.
 * @returns characteristics for the requested level.
 */
function describeSeverity(severity: NormalizedSeverity): SeverityAttributes {
  const attributes = severityAttributesLookup.get(severity);
  if (attributes === undefined) {
    throw new Error('Unhandled severity level');
  }
  return attributes;
}

/**
 * Helper that throws when the compiler believes an exhaustive switch is not completed.
 * @param _value value that should never be reachable.
 */
function assertNever(_value: never): never {
  throw new Error('Unhandled severity level');
}

/**
 * Forward a structured JSON line to the Tauri logging plugin when available.
 * @param line serialized log line to ship.
 */
async function forwardToTauri(line: string): Promise<void> {
  try {
    const log = await import('@tauri-apps/plugin-log');
    await log.info(line);
  } catch {
    // ignore missing plugin
  }
}

/**
 * Emit a structured log entry to both console and the Tauri plugin.
 * @param entry log payload provided by callers.
 */
export async function logMessage(entry: LogEntry): Promise<void> {
  const context = await getLoggingContext();
  const severity = normalizeSeverity(entry.severity);
  const severityAttributes = describeSeverity(severity);

  const record = {
    timestamp: new Date().toISOString(),
    level: severityAttributes.level,
    'syslog.severity': severityAttributes.severityNumber,
    'syslog.severity_text': severityAttributes.severityText,
    message: entry.message,
    component: entry.component,
    event_name: entry.eventName,
    correlation_id: entry.correlationId ?? 'unknown',
    session_id: context.sessionId,
    build: {
      version: context.buildVersion,
      commit: context.buildCommit ?? 'unknown',
    },
    platform: {
      os: context.platformOs,
      arch: context.platformArch,
    },
    source: {
      layer: 'webview',
      module: entry.source?.module ?? 'unknown',
      function: entry.source?.function ?? 'unknown',
    },
    ...entry.metadata,
  };

  const line = JSON.stringify(record);
  const consoleReference = (globalThis as typeof globalThis & { console: Console }).console;
  const formattedLine = `[${severity}] ${line}`;
  switch (severityAttributes.consoleMethod) {
    case 'error': {
      consoleReference.error(formattedLine);
      break;
    }
    case 'warn': {
      consoleReference.warn(formattedLine);
      break;
    }
    case 'info': {
      consoleReference.info(formattedLine);
      break;
    }
    case 'debug': {
      consoleReference.debug(formattedLine);
      break;
    }
    default: {
      assertNever(severityAttributes.consoleMethod);
    }
  }
  await forwardToTauri(line);
}
